### PR TITLE
Rewrite _get_total_norm with _powsum, add skip_root flag

### PR DESCRIPTION
**Pull request by pianpwk**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* #172620
* __->__ #172603

For gradient clipping, the _get_total_norm function uses a 2-stage linalg.vector_norm reduction, once per parameter, then once across all parameter norms. For sharded DTensors this incurs a sqrt -> pow -> sqrt loop which degrades numerics.

This rewrites it for p-norms, to use _powsum instead, incurring a single-scalar allreduce, and a single sqrt at the end, for better numerics. For norms 0/1/inf/-inf there is no loop, so behavior is unchanged.

Also adds a skip_root flag defaulting to False, so cross-mesh reductions can also delay the root.

Authored with Claude.

### CODE REVIEW

File: torch/nn/utils/clip_grad.py [hunk 1]
```diff
@@ -89,23 +89,42 @@ def _get_total_norm(
     )  # type: ignore[assignment]
 
     norms: list[Tensor] = []
+    # For p-norms (p not in {inf, -inf, 0, 1}), use powsum to accumulate sum(|x|^p)
+    # and apply the root once at the end. This avoids sqrt -> pow -> sqrt cycles.
+    use_powsum = norm_type not in (float("inf"), float("-inf"), 0.0, 1.0)
```
Comment (ezyang) [inline torch/nn/utils/clip_grad.py]: Can you tell me why the non p-norm cases powsum doesn't work?

Comment (pianpwk) [reply torch/nn/utils/clip_grad.py]: I removed the special-casing for p=1, but for p=inf/-inf/0, the decomposition doesn't really hold
- norm(x, p=0) is count(x != 0), but sum(x**0) already clobbers over that info
- norm(x, p=inf and -inf) are max and min of abs(x), but sum(abs(x)**(inf or -inf)) also clobber over this

I guess we could change semantics so powsum for these norms return the same thing as vector_norm, and just skip the root. Would that be more appropriate/less surprising?

Comment (ngimel) [reply torch/nn/utils/clip_grad.py]: "skip the root" for `**inf` doesn't really make sense, I think it's fine to skip those cases

**Comment by wconstab** [comment]
@claude review this PR

File: torch/nn/utils/clip_grad.py [hunk 2]
```diff
@@ -88,23 +93,44 @@ def _get_total_norm(
     )  # type: ignore[assignment]
 
     norms: list[Tensor] = []
+    # For p-norms (p not in {inf, -inf, 0}), use powsum to accumulate sum(|x|^p)
+    # and apply the root once at the end. This avoids sqrt -> pow -> sqrt cycles.
+    use_powsum = norm_type not in (float("inf"), float("-inf"), 0.0)
```
Comment (wconstab) [inline torch/nn/utils/clip_grad.py:98]: @claude can you suggest a brief docstring addition for this line that explains why we have these exceptions (taking the other PR comments from ed, pian and natalia into account)

File: torch/nn/utils/clip_grad.py [hunk 3]
```diff
@@ -88,23 +93,44 @@ def _get_total_norm(
...
+    stacked_norms = torch.stack([norm.to(first_device) for norm in norms])
+
+    if use_powsum:
+        # norms are sum(|x|^p), just sum them
+        total_norm = stacked_norms.sum()
+        if not skip_root:
+            total_norm = total_norm ** (1.0 / norm_type)
+    else:
+        # For inf/-inf/0 norms, use vector_norm directly
+        total_norm = torch.linalg.vector_norm(stacked_norms, norm_type)
```
Comment (wconstab) [inline torch/nn/utils/clip_grad.py:133]: agree w/ claude- we are missing checking skip_root in this path, and should be loud if its unsupported

File: torch/nn/utils/clip_grad.py [hunk 4]
```diff
@@ -51,6 +51,7 @@ def _get_total_norm(
     norm_type: float = 2.0,
     error_if_nonfinite: bool = False,
     foreach: bool | None = None,
+    skip_root: bool = False,
```
Comment (ezyang) [inline torch/nn/utils/clip_grad.py:54]: I understand this is a private function, but I'm not super enthusiastic about this as a general API concept. Perhaps would be even better as a private kwarg even here, if we absolutely wanted to do it this way.

**Review by ezyang — CHANGES_REQUESTED** [review]
don't do this on its own. Squash with the use site
